Split `microovn.central` service.

mkalcok opened the pull request:
Previously `microovn.central` started all three central services (ovn-nb, ovn-sb and northd). Now these daemons are split into their own services for more fine-grained controll over them.

The `microovn.central` will stay for now and act as an umbrella service that can trigger start of all three services. However the usage is discouraged as it will be removed. It's also not capable of being used as shortcut for stopping these services.

Additional consequence of splitting of the services is that we can now scale the cluster without restarting OVN NB/SB services as only the northd daemon needs restart.

fnordahl commented on snapcraft/ovn-central.env:
```diff
@@ -0,0 +1,21 @@
+export OVN_SYSCONFDIR="${SNAP}/etc"
+export OVN_PKIDIR="${SNAP_COMMON}/data/pki"
+
+# Disable some commands
+mkdir -p "${OVN_RUNDIR}/bin/"
+for i in install plymouth sudo systemctl; do
+    [ -e "${OVN_RUNDIR}/bin/${i}" ] && continue
+
+    ln -s "/bin/true" "${OVN_RUNDIR}/bin/${i}"
+done
```
While I understand how these ended up in this file, it's probably not the correct place to host them after the split.

Having three services starting independently attempting to create this might lead to a race conditions and spurious errors. (And having an sourceable "environment" file performing actions behind your back comes as an element of surprise.)

Would it make sense to move this to either the install hook or as part of the startup of the `daemon` service?

fnordahl replied:
I'd rather not have them in a sourceable `.env` file.

Speaking of who needs this, I honestly do not think this is related to any of the central services at all, I do not find any evidence of the `ovn-ctl`, `ovs-ctl` or `ovs-lib` scripts making calls to these binaries. So I wonder what was the reason for the original introduction of this, habits of the original author?

mkalcok replied:
I'd be in favor of removing this then. The symlinks do not even have effect without the very specific `{$OVN_RUNDIR}/bin/` added to path, and that happens only here.

fnordahl commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+
+    Deprecated: This service will be removed in future release
+
+This is a shortcut service. Starting this service will start and enable
```
s/shortcut/transitional

pmatulis replied:
... and enable multiple services:

* ``microovn.ovn-nb``
* ``microovn.ovn-sb``
* ``microovn.ovn-northd``

However, ...

fnordahl commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+This is a shortcut service. Starting this service will start and enable
+``microovn.ovn-nb``, ``microovn.ovn-sb`` and ``microovn.ovn-northd``. However
+this service , due to the ``snap`` limitations, is not capable of stopping
+child services and its usage is generally discouraged.
```
I'd rather highlight the reasons for why managing these services separately is desirable rather than bashing snaps here

fnordahl replied:
Sure, I'd angle it with a message telling the user to use the other services to stop things instead then. Together with the transitional label it will make it absolutely clear to the user that they should stop whatever it is they think they should be doing and follow the new workflow

fnordahl commented on snapcraft/commands/central.start:
```diff
@@ -1,67 +1,15 @@
-    --ovn-manage-ovsdb=no --no-monitor
-
-sleep infinity
+echo "This service can be used as a shortcut to start ovn-nb, ovn-sb and \
```
s/shortcut/transitional

tomponline reviewed (commented):
I'm looking forward to this landing :)

mkalcok commented on microovn/ovn/bootstrap.go:
```diff
@@ -68,9 +68,9 @@ func Bootstrap(s *state.State) error {
-        if err != nil {
-                return fmt.Errorf("failed to generate TLS certificate for client: %s", err)
-        }
+	if err != nil {
+		return fmt.Errorf("failed to generate TLS certificate for client: %s", err)
+	}
```
MicroOVN seems to be using tabs for indentation and this piece of code used spaces.This was an auto-formater change and I'd leave it in for consistency.

fnordahl replied:
Cleaning up inconsistencies is always a nice thing, but perhaps they belong in a separate commit?

mkalcok commented on microovn/ovn/bootstrap.go:
```diff
@@ -93,9 +93,19 @@ func Bootstrap(s *state.State) error {
 
 	// Enable OVN central.
-	err = snapStart("central", true)
+	err = snapStart("ovn-nb", true)
```
Ideally users should not interact with these services at all...but I see your point and I don't see any downside to this change, aside from bit of verbosity

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+=================
+MicroOVN Services
```
s/Services/services

pmatulis commented on docs/reference/index.rst:
```diff
@@ -0,0 +1,11 @@
+
+These references describe concepts unique to MicroOVN. They do not aim to
+describe general OVN/OVS concepts
+
```
MicroOVN reference material is specific to the MicroOVN project. It does not cover upstream OVN/OVS topics.

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+This document provides brief explanation of services provided by MicroOVN. They
+are generally not meant for direct user interaction and should be used only
+if necessary. List and status of these services can be displayed by running:
+
```
This page presents a list of all MicroOVN services. The descriptions presented here is for reference only - the user is not expected to interact directly with these services.

The status of all services is displayed by running:

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+
+.. code-block:: none
+
+    snap services microovn
```
Align the first letter of the command ('s') directly under the 'c' (of code-block). This should be done everywhere in the document, even for 'note' or 'warning', etc.

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+
+.. warning::
+
+    Deprecated: This service will be removed in future release
```
The ``microovn.central`` service is deprecated and will be removed in a future release.

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+
+    snap services microovn
+
+microovn.central
```
Let's put this in monospace (double backticks) - same for the rest of the headers with service names.

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+microovn.chassis
+----------------
+
+This service maps directly to ``ovn-controller`` daemon.
```
... the ``ovn-controller``

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+microovn.daemon
+---------------
+
+Main MicroOVN service/process that manages rest of the processes. It also
```
The main ...

... manages all the other processes.

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+
+Main MicroOVN service/process that manages rest of the processes. It also
+handles communication with other MicroOVN cluster members and provides an API
+for ``microovn`` client command.
```
... for the ``microovn`` ...

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+microovn.ovn-nb
+---------------
+
+This service maps directly to ``OVN Northbound`` database/service.
```
... directly to the ...

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+microovn.ovn-northd
+-------------------
+
+This service maps directly to ``ovn-northd`` daemon.
```
... directly to the ...

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+microovn.ovn-sb
+---------------
+
+This service maps directly to ``OVN Southbound`` database/service.
```
... directly to the ...

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+
+This service is a recurring process that runs once a day between ``02:00`` and
+``02:30``. It triggers TLS certification reissue for certificates that are
+nearing the expiration. For more information see ``Certificate lifecycle``
```
... information see the ...

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,63 @@
+This service is a recurring process that runs once a day between ``02:00`` and
+``02:30``. It triggers TLS certification reissue for certificates that are
+nearing the expiration. For more information see ``Certificate lifecycle``
+section in :doc:`/how-to/tls`
```
Link directly to that section. See:

https://docs.openstack.org/charm-guide/latest/community/doc-contrib/rst.html#linking-to-a-location-within-a-page-in-the-doc-set

pmatulis reviewed (changes requested):
Thanks for this. See change requests.

I would normally expect to see a significant doc change like this as a separate PR with a separate Jira card.

mkalcok commented on docs/reference/index.rst:
```diff
@@ -0,0 +1,11 @@
+==========
+References
```
I followed naming convention in other documentation modes. Both `How-to guides` and `Tutorials` use plural.

mkalcok commented:
Thank you for extensive review @fnordahl @pmatulis . I believe I addressed all the comments.

> I would normally expect to see a significant doc change like this as a separate PR

The need for this documentation was created directly by code changes in this PR. We split the `microovn.central` service but retained it (for some underlying technical reasons) which, imo, created a bit of an ambiguity. That's why we decided to include also this documentation piece that would explain the purpose of each service exposed by the MicroOVN snap.

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,71 @@
+
+* ``microovn.ovn-ovsdb-server-nb``
+* ``microovn.ovn-ovsdb-server-sb``
+* ``microovn.ovn-northd``.
```
extra dot

pmatulis commented on docs/reference/services.rst:
```diff
@@ -0,0 +1,71 @@
+MicroOVN services
+=================
+
+This page presents a list of all MicroOVN services. The descriptions presented
```
apologies, my fault here:

s/The descriptions presented here is/Their descriptions are/

pmatulis reviewed (changes requested):
Sorry, a few last things

fnordahl commented on snapcraft/commands/central.start:
```diff
@@ -1,67 +1,15 @@
-export OVN_SYSCONFDIR="${SNAP}/etc"
-export OVN_PKIDIR="${SNAP_COMMON}/data/pki"
-
-# Disable some commands
-mkdir -p "${OVN_RUNDIR}/bin/"
-for i in install plymouth sudo systemctl; do
-    [ -e "${OVN_RUNDIR}/bin/${i}" ] && continue
-
-    ln -s "/bin/true" "${OVN_RUNDIR}/bin/${i}"
```
Good riddance!

However, I see these are also in `snapcraft/commands/chassis.start` and `snapcraft/commands/switch.start`.

Would it be more consistent to drop all of these simultaneously in a separate parent commit?


For the record, there are now some additional `DENIED` messages logged, but they do not affect operation.
```
[  101.595042] audit: type=1400 audit(1697770466.057:241): apparmor="DENIED" operation="exec" profile="snap.microovn.ovn-ovsdb-server-nb" name="/usr/bin/systemctl" pid=2804 comm="ovn-ctl" requested_mask="x" denied_mask="x" fsuid=0 ouid=0
[  101.595086] audit: type=1400 audit(1697770466.057:242): apparmor="DENIED" operation="exec" profile="snap.microovn.ovn-ovsdb-server-nb" name="/usr/bin/systemctl" pid=2804 comm="ovn-ctl" requested_mask="x" denied_mask="x" fsuid=0 ouid=0
[  101.596966] audit: type=1400 audit(1697770466.061:243): apparmor="DENIED" operation="exec" profile="snap.microovn.ovn-ovsdb-server-nb" name="/usr/bin/plymouth" pid=2806 comm="ovn-ctl" requested_mask="x" denied_mask="x" fsuid=0 ouid=0
[  101.597023] audit: type=1400 audit(1697770466.061:244): apparmor="DENIED" operation="exec" profile="snap.microovn.ovn-ovsdb-server-nb" name="/usr/bin/plymouth" pid=2806 comm="ovn-ctl" requested_mask="x" denied_mask="x" fsuid=0 ouid=0
```

Even without this change there are still other unhandled `DENIED` rules logged, such as:
```
[  101.599507] audit: type=1400 audit(1697770466.065:245): apparmor="DENIED" operation="capable" profile="snap.microovn.switch" pid=2787 comm="ovsdb-server" capability=38  capname="perfmon"
```

So I still think we should still drop these hacks and figure out what indirect dependencies of the OVS/OVN ctl scripts end up calling the offending binaries and provide patches upstream at some point in the future.

mkalcok replied:
I agree. I'll drop those instances too. The commands do not work anyway, it's better to have the failures logged rather than silently failing.

fnordahl commented on docs/how-to/tls.rst:
```diff
@@ -112,6 +112,8 @@ member.
    member. Any failure will result in subsequent communication errors for that
    service within the cluster.
 
+.. _certificates_lifecycle:
```
Move the doc changes to a separate commit? Some of these changes will surely cause conflicts on backport, and it would be very nice to not have to deal with conflicts on the actual code changes.

mkalcok replied:
Yeah. I didn't slice this PR into commits very well.

fnordahl reviewed (changes requested):
I took this for a spin and tested both refresh, fresh install and post configuration reboots, looks great.

The commit messages may need updating to reflect changes during review.

I also have a couple of comments in-line that request splitting a couple of things into separate commits.

mkalcok commented:
Thanks again for the review. I incorporated your comments and I separated the docs and the command masking removal into separate commits.

fnordahl reviewed (approved):
LGTM, thanks!
